### PR TITLE
[sertop] Update to sertop upstream.

### PR DESCRIPTION
**Pull request by ejgallego**
A rather hacky patch to make elcoq work again with current sertop/Coq master.

Note the disappearance of `ErrMsg` in favor of `Message` with level `Error`, also, most commands such as `Query` of `StmAdd` take options with named records now, assuming a default if no option given.

### CODE REVIEW

File: elcoq.el [hunk 1]
```diff
@@ -247,11 +247,16 @@ are fine to ignore."
...
             ;; `details' has `fname', `line_nb', `bol_pos', `line_nb_last',
             ;; `bol_pos_last', `bp', `ep'.
-            (elcoq--sertop-set-overlay-status overlay 'Failed)
-            (overlay-put overlay 'after-string (format "\n%s\n" message)))
+            (pcase level
+              (`Error
+               (elcoq--sertop-set-overlay-status overlay 'Failed)
+               (overlay-put overlay 'after-string (format "\n%s\n" message)))
+              (other
+               ;; Warning, Notice, Info, Debug
```
Comment (cpitclaudel) [inline elcoq.el:263]: Let's have a message here :)

File: elcoq.el [hunk 2]
```diff
@@ -476,18 +481,19 @@ sertop running."
...
                 (overlay-end overlay))))
-      `(Control (StmAdd
-                 1 ;; Just one sentence for now
-                 ,(if previous-state `(Some ,previous-state) `None)
-                 ,str)))))
+      ;; If no options given, StmAdd will try "the rigth thing"
+      `(Control (StmAdd () ,str)))))
+                 ;; 1 ;; Just one sentence for now
+                 ;; ,(if previous-state `(Some ,previous-state) `None)
+                 ;; ,str)))))
```
Comment (cpitclaudel) [inline elcoq.el]: This doesn't look good: elcoq doesn't keep track of what state is current, so we'd much rather have an add on the wrond state id rejected than have it silently accepted and added somewhere else in the document, right?

Comment (cpitclaudel) [inline elcoq.el]: Also, typo in `rigth`

**Comment by cpitclaudel** [comment]
LGTM with one main comment: I don't think SerAPI doing the "right thing" is a good idea. If my document is 

```
Goal A. Proof. a. b. c.  <1> Qed.
Goal B. Proof. a. b. <2>
```

then there are two points where I can legitimately make edits and add sentences, namely `<1> and`<2>`. IIUC currently the Stm will in fact only let me add to one of them (I can't jump back and forth between adding to one and adding to the other) ­– is that right?

In that case, SerAPI should force me to tell it which state id I think I'm appending to, and yell at me if I append to the wrong one. Doing the "right thing" will cause SerAPI to build a model of the document with sentences attached in the wrong places, which will be confusing (and hard to debug!).

IOW, I think this is a case of explicit beats implicit :) What do you think?

**Comment by ejgallego** [comment]
Hi Clément, thanks for the comments; PR updated.

I fully agree with you in what the "right thing" should be.

Now, we have two notions for the "right" thing, Coq's internal one,
and ours, the external one. And as it is common, they both disagree.
- From the point of view of the STM, there is a single "tip" on top of
  which you can add, so it is pointless to specify an `ontop`
  parameter. See the code:

``` ocaml
  let cur_tip = VCS.cur_tip () in
  if not (Stateid.equal ontop cur_tip) then
    (* For now, arbitrary edits should be announced with edit_at *)
    anomaly(str"Not yet implemented, the GUI should not try this");
```

  This is a bit inconvenient but we will have to live with it for a
  while.
- From our point of view, we have many tips corresponding to
  proofs. However, it is tricky to go from one to another due to
  (mainly parsing) side effects.

The idea of the optional parameter indeed is to handle the initial state as you can see now. Indeed, in jsCoq I just can pass `null` and it will work. However, note that SerAPI is going to declare `1` as initial state officially now. :D

**Pull request closed without merging**